mage_integrations/mage_integrations/tests/sources/test_base.py

```diff
@@ -0,0 +1,528 @@
+                ],
+                'metadata': {
+                    'inclusion': 'available',
+                    'selected': True,
+                }
+            },
+        ],
+        replication_method='INCREMENTAL',
+        schema=dict(
+            properties=dict(
+                age=dict(type=['null', 'integer']),
+                id=dict(type=['null', 'string']),
+                last_name=dict(type=['null', 'string']),
+                first_name=dict(type=['null', 'string']),
+                color=dict(type=['null', 'string'])
+            ),
+            type='object',
+        ),
+        stream='demo_users',
+        tap_stream_id='demo_users',
+        unique_conflict_method='UPDATE',
+    )
+
+    return [stream1, stream2]
+
+def build_sample_streams_catalog():
+    return Catalog.from_dict(
+        dict(streams=build_sample_streams_list())
+    )
+
+class BaseSourceTests(unittest.TestCase):
+    maxDiff = None
+
+    def test_templates(self):
+        # Template folders exist in the different integration source folders.
+        source = PostgreSQL()
+        templates = source.templates()
+        self.assertEqual(
+            templates,
+            {
+                'config': {
+                    'database': '',
+                    'host': '',
+                    'password': '',
+                    'port': 5432,
+                    'schema': '',
+                    'username': '',
+                    'replication_slot': '',
+                }
+            },
+        )
+
+    def test_discover(self):
+        # Testing with Intercom source, since it has a "schemas" folder and no "discover" subclass method.
+        source = Intercom()
+        streams = source.discover().streams
+        self.assertEqual(len(streams), 11)
+
+    def test_discover_streams(self):
+        source = Source()
+        with patch.object(source, 'get_stream_ids', return_value=['demo_table', 'demo_users']):
+            streams = source.discover_streams()
+            self.assertEqual(
+                streams,
+                [
+                    dict(
+                        stream='demo_table',
+                        tap_stream_id='demo_table',
+                    ),
+                    dict(
+                        stream='demo_users',
+                        tap_stream_id='demo_users',
+                    ),
+                ],
+            )
+
+    def test_get_stream_ids(self):
+        source = Source()
+        with patch.object(source, 'discover', return_value=build_sample_streams_catalog()):
+            stream_ids = source.get_stream_ids()
+            self.assertEqual(stream_ids, ['demo_table', 'demo_users'])
+
+    def test_process_execute_test_connection(self):
+        source = Source(
+            test_connection=True,
+        )
+        with patch.object(source, 'test_connection') as mock_test_connection:
+            source.process()
+            mock_test_connection.assert_called_once()
```

Comment:
let's also assert some other methods not called. same for other conditions of the `process` method.